### DISCUSSION TITLE
Control plane availability

### DISCUSSION THREAD

**Discussion by somehowchris** [karmada-io/karmada#3773] [Q&A]
In general I really like karmada but there's 1 bugging me.

Is the karmada control plane itself HA?
-> generally having a small outage over a few minutes/hours isn't really bad [interms of control of the cluster ] but as more and more things are managed via karmada, how does one keep up with leader election in pull mode? is there even something like leader election? also in the case of taking down a node i.e. migrating between zones, how does one demote an old host cluster to have another one (preferably automatically) be prompoted?

**Comment by whitewindmills** [comment]
it's a good question, but currently Karmada control plane does not support HA. we are also considering the HA solution for the Karmada control plane.

**Comment by RainbowMango** [comment]
From the [Karmada users](https://karmada.io/docs/next/casestudies/adopters) practice, Karmada control plane HA could be achieved in the following two ways:
1. All Karmada components could be run with leader election mode, which means people can run multiple replicas for each component, just like the practice of how we run Kubernetes in HA mode. This is a common way.
2. Run multiple Karmada components, a member cluster might be registered to more than one Karmada control plane, so even in case of one Karmada downs, people can still manage the cluster as well as the resources by another Karmada instance. This approach is still less practiced, but it does exist.

**Reply by somehowchris** [reply]
Thanks for the swift reply.

Interesting approaches. I have read about the leader election flags but didn't have any further context, is there a page or something like a blog explaining a bit more? This for sure would be lovely to read and tinker more about.

About the second approach, so to my understanding, karmada only propagates ressources which it is told to share like deployments or pods. So with that in mind how would other karamda clusters have the same context? Also there is the FedereatedHPA, how would one keep track which karmada host cluster schedules which scaling?

**Reply by RainbowMango** [reply]
Unfortunately, Karmada doesn't have specific documents for HA yet in addition to [general installation](https://karmada.io/docs/installation/). Thanks for spotting this, I just [opened an issue](https://github.com/karmada-io/website/issues/418) to track this. 

Before the official document, i can briefly share some ideas about how to achieve HA for Karmada.
If you have experience deploying HA Kubernetes cluster, then deploying HA Karmada will be very simple and similar:
- karmada-apiserver: it's essentially a `kube-apiserver`.
- karmada-controller-manager: similar to `kube-controller-manager`.
- karmada-scheduler: similar to `kube-scheduler`.
- so on
The working mechanism of Karmada components is similar to Kubernetes, I don't think there is anything would block Karmada work in HA mode.

The second approach may be a higher level of HA and may exceed the scope of this conversation. On the other hand, I will also gather feedback from users and hope to explain this practice through an official doc.

**Reply by cmontemuino** [reply]
We're piloting Karmada to use it as a regional control plane and the HA capability of Karmada components makes sense to us when having stretched kubernetes clusters.
The key question here is what happens when you have a Kubernetes cluster in a single zone, with Karmada configured in HA, but suddenly your K8s cluster is not available, or the AZ is not available.

For such a case, we're exploring to have two Karmada CPs, each one in a separate AZ, in an Active-Standby fashion.

When a Karmada CP goes down (for whatever reason: non accessible, or just not operational), then we "cordon" that CP and activate the one which is in Standby mode.
- "Standby" is everything operational, but we short-circuit the communication to all member clusters

Besides this approach is tackling the availability concern, we think it might also help when upgrading K8s (i.e., the host cluster) and/or Karmada itself.

Just wanted so share in case it sparks some further ideas.

**Reply by RainbowMango** [reply]
> For such a case, we're exploring to have two Karmada CPs, each one in a separate AZ, in an Active-Standby fashion.

Having a member cluster managed by two separate Karmada instances would be chanllenging. The main issue is that each karmada instance creates credentials on the member cluster, which can lead to conflicts and unpredictable behavior. 

However, I've seen an interesting approach from one of our users where they achieved high availability across regions differently. They deployed **two sets of Karmada control plane replicas in different regions, but had them share the same ETCD cluster**. This shared storage layer allows both control plane instances to maintain consistency while achieving cross-region resilience.

@jabellard  if you're aware of any public talks or blog posts about this deployment pattern, please feel free to share them here!

**Reply by cmontemuino** [reply]
We're also looking forward to have an external etcd. That's going to be the very next step after getting Karmada v1.17.0 (hopefully with etcd encryption enabled). This will effectively move the single point of failure from Karmada to etcd, but still, it's an interesting option.

~What the external etcd does not solve is the "switch-over" mechanism, which we want to use in two different situations~:
- ~Upgrading Kubernetes in the host cluster~ <-- that's ok
- ~Upgrading Karmada itself~ <-- also ok

Only caveat is etcd becoming the problem (etcd upgrade, k8s upgrade, etc.). So, from Karmada POV that should be ok.

Regarding the dual-KCP approach, we're testing it as of this writing. So far so good. We didn't observe issues at all, at least in Push mode. The following remark is very relevant:
> The main issue is that each karmada instance creates credentials on the member cluster, which can lead to conflicts and unpredictable behavior.

That's a limitation of `karmadactl`, which does not allow to configure much of where to store secrets, name of the ClusterRole for `cluster-admin`, etc. We had to create a custom script to create and maintain all the credentials completely isolated.

A blog post is coming out of this effort.

**Reply by cmontemuino** [reply]
Following up on this discussion — I've been reviewing the implementation of `pkg/controllers/cluster/cluster_controller.go, tryUpdateClusterHealth` and `pkg/controllers/status/cluster_status_controller.go, syncClusterStatus`:
- https://github.com/karmada-io/karmada/blob/v1.16.2/pkg/controllers/cluster/cluster_controller.go#L452
- https://github.com/karmada-io/karmada/blob/v1.16.2/pkg/controllers/status/cluster_status_controller.go#L181

If we want to keep our standby pattern with a **shared etcd**, then it would create a continuous write conflict between active KCP and standby one. The former will write `Ready=True` (it can reach the member), while the standby's writes `Ready=False`, both retrying on conflict via `retry.RetryOnConflict`. The active KCP's `cluster-controller` would then oscillate taints (`UnreachableTainTemplateForSched` on/off), potentially disrupting scheduling.

Question (cc @RainbowMango): Is this analysis correct? 
I wonder if the approach is using a shared etcd to have one single KCP at a time, and bringing a newer one on board when detecting a failure?
Or is the only safe approach with shared etcd to keep the secondary KCP fully stopped (components scaled to 0)?

Another option in mind is that you're thinking about having one single KCP at a time. If that's the case, then we need to make sure Karmada doesn't touch the `karmada-cluster` namespace (where there are secrets with the tokens). Otherwise, the new KCP won't be able to connect to the member clusters and (I think) there's no other way around other than re-register every member cluster.

*Please note I'm targeting Push mode*.

Thanks in advance — this would be very helpful for understanding the limits of shared-etcd approach.

**Reply by jabellard** [reply]
Hey everyone! This is a great discussion. At Bloomberg, we built a large scale and fully managed multi-cluster platform built atop the Karmada project to power modern AI and streaming analytics workloads. I talked bout how we solved this HA problem at KubeCon EU last year. Check out [this talk](https://www.youtube.com/watch?v=rbVV8WIJYww&t=298s) to learn more.

**Reply by RainbowMango** [reply]
> I wonder if the approach is using a shared etcd to have one single KCP at a time, and bringing a newer one on board when detecting a failure?
Or is the only safe approach with shared etcd to keep the secondary KCP fully stopped (components scaled to 0)?

If two Karmada instances share the same etcd cluster, their controller instances can run in leader election mode—ensuring that only one controller instance is active at any given time, thereby avoiding any conflicts.
For example, if the active Karmada instance currently holds the leader election resource lock, the standby Karmada instance remains idle until the active instance goes down.

I have a diagram here that illustrates this idea—with a slight difference: it uses a PostgreSQL database instead of an etcd cluster.
<img width="5380" height="2973" alt="image" src="https://github.com/user-attachments/assets/96da2b72-b6dc-4d13-9b9b-061c9f22fc15" />

**Reply by jabellard** [reply]
@RainbowMango , I find that one common use case of Karmada is to leverage multi-cluster orchestration to achieve redundancy across multi geographic regions. That usually requires that the orchestration layer (i.e., Karmada control plane) itself be setup in HA mode. This is how we set things up for our tenants at Bloomberg. Do you think a blog post/ admin guide to provide general guidance for users for how to deploy Karmada control planes in HA mode would be useful to the wider community? If so, I can write it. Just let me know what you think.

**Reply by RainbowMango** [reply]
Sure, that would be very helpful! Will this be included in the case study blog as well?
